python/cugraph/dask/pagerank/pagerank.py

```diff
@@ -0,0 +1,253 @@
+    """
+    dev, gpu_array, idx = data
+
+    in_handle_src = gpu_array[0].get_ipc_handle()
+    in_handle_dest = gpu_array[1].get_ipc_handle()
+    return dev, [in_handle_src, in_handle_dest], idx
+
+
+def _build_host_dict(gpu_futures, client):
+    """
+    Build a dictionary of hosts and their corresponding ports from workers
+    which have the given gpu_futures.
+    """
+    # TO DO: IMPROVE/ CLEANUP
+    who_has = client.who_has(gpu_futures)
+
+    workers = [key[0] for key in list(who_has.values())]
+    hosts = set(map(lambda x: parse_host_port(x), workers))
+    hosts_dict = {}
+    for host, port in hosts:
+        if host not in hosts_dict:
+            hosts_dict[host] = set([port])
+        else:
+            hosts_dict[host].add(port)
+
+    return hosts_dict
+
+
+def _mg_pagerank(data):
+    """
+    Collect all ipc pointer information into source and destination alloc_info
+    list that is passed to snmg pagerank.
+    """
+    ipcs, raw_arrs, alpha, max_iter = data
+
+    # Separate threads to hold pointers to separate devices
+    # The order in which we pass the list of IPCs to the thread matters and
+    # the goal isto maximize reuse while minimizing the number of threads.
+    # We want to limit the number of threads to O(len(devices)) and want to
+    # avoid having if be O(len(ipcs)) at all costs!
+    device_handle_map = defaultdict(list)
+    [device_handle_map[dev].append((idx, ipc)) for dev, ipc, idx in ipcs]
+
+    open_ipcs = []
+    for dev, ipcs in device_handle_map.items():
+        open_ipcs.append([[dev], new_ipc_thread(ipcs[0][1], dev)])
+
+    alloc_info_src = []
+    alloc_info_dest = []
+    for dev, t in open_ipcs:
+        inf = t.info()
+        for i in range(len(dev)):
+            alloc_info_src.append([get_device_id(dev[i]), inf[0]])
+            alloc_info_dest.append([get_device_id(dev[i]), inf[1]])
+
+    for t in raw_arrs:
+        raw_info = build_alloc_info(t)
+        alloc_info_src.append([get_device_id(t[2]), raw_info[0]])
+        alloc_info_dest.append([get_device_id(t[2]), raw_info[1]])
+
+    alloc_info_src.sort(key=lambda x: x[0])
+    alloc_info_dest.sort(key=lambda x: x[0])
+
+    final_allocs_src = [a for i, a in alloc_info_src]
+    final_allocs_dest = [a for i, a in alloc_info_dest]
+
+    pr = cugraph.mg_pagerank(final_allocs_src,
+                             final_allocs_dest,
+                             alpha, max_iter)
+
+    [t[1].close() for t in open_ipcs]
+    [t[1].join() for t in open_ipcs]
+
+    return pr
+
+
+def pagerank(edge_list, alpha=0.85, max_iter=30):
+    """
+    Find the PageRank values for each vertex in a graph using multiple GPUs.
+    cuGraph computes an approximation of the Pagerank using the power method.
+    The input edge list should be provided in dask-cudf dataframe
+    with one partition per GPU.
+
+    Parameters
+    ----------
+    edge_list : dask_cudf.DataFrame
+       Contain the connectivity information as an edge list
```

Comment:
Missing period at the end. And it seems like using 4 spaces is most common in cugraph, cudf, & cuml docstrings (here, 3 spaces are used).

Reply:
fixed